### PR TITLE
added descriptive data assess

### PR DESCRIPTION
**Pull request by amrasekh**
Added descriptive dataset assessment

- distribution of samples across groups
- distribution of labels across groups
- statistical parity and disparate impact for all preferred label value possibilities

### CODE REVIEW

**Review by IanAtCredo — CHANGES_REQUESTED** [review]

File: credoai/modules/dataset_modules/dataset_base.py [hunk 1]
```diff
@@ -155,7 +157,60 @@ def _calculate_mutual_information(self, categorical_features=None, normalize=Tru
                 results[k] = {'value':v, 'feature_type': 'continuous'}
 
         return results
-    
 
+    def _concat_features_label(self):
```
[the remark this answers is not in the document]
Comment (amrasekh) [reply credoai/modules/dataset_modules/dataset_base.py]: Good idea. It is now moved to `CredoData`.

File: credoai/modules/dataset_modules/dataset_base.py [hunk 2]
```diff
@@ -155,7 +157,60 @@ def _calculate_mutual_information(self, categorical_features=None, normalize=Tru
...
+        results['label_balance'] = lb
+
+        # Fairness metrics
+        r = df.groupby([sensitive_feature_name, label_name]).agg({label_name: 'count'})
+        r = r.groupby(level=0).apply(lambda x: 100 * x / float(x.sum()))
+        r.rename({label_name:'ratio'}, inplace=True, axis=1)
+        r.reset_index(inplace=True)
+        sp = r.groupby(label_name)['ratio'].apply(lambda x: np.max(x)-np.min(x)).reset_index(name='value').to_dict(orient='records')
+        di = r.groupby(label_name)['ratio'].apply(lambda x: np.max(x)/np.min(x)).reset_index(name='value').to_dict(orient='records')
+        results['metrics'] = {'maximum_statistical_parity': sp, 'maximum_disparate_impact': di}
```
[the remark this answers is not in the document]
Comment (amrasekh) [reply credoai/modules/dataset_modules/dataset_base.py]: Good idea. I updated the naming.
It looks very related, but I could not think of a straightforward way now to configure it to compute the metric of interest and for all the preferred label possibilities. Also seems it requires setting `y_pred`, but for this one a dummy one can be passed but have the metric function to ignore it. Seems doable but would take some time digging into it.

File: credoai/modules/dataset_modules/dataset_base.py [hunk 3]
```diff
@@ -156,6 +159,44 @@ def _calculate_mutual_information(self, categorical_features=None, normalize=Tru
...
 
-    
+        Returns
+        -------
+        dict
+            'sample_balance': distribution of samples across groups
+            'label_balance': distribution of labels across groups
+            'metrics': maximum statistical parity and maximum disparate impact between groups for all preferred label value possibilities 
+        """
+        credo_data = cl.CredoData(name='dataset',
```
Comment (IanAtCredo) [inline credoai/modules/dataset_modules/dataset_base.py:172]: Oh, this is weird. CredoData should not be made within a module. I would redefine `_concat_features_label_to_dataframe` as [static method](https://www.geeksforgeeks.org/class-method-vs-static-method-python/). You can then use it here, as well as within CredoData in the future if we need.

File: credoai/modules/dataset_modules/dataset_base.py [hunk 4]
```diff
@@ -156,6 +159,44 @@ def _calculate_mutual_information(self, categorical_features=None, normalize=Tru
...
+                          X=self.X,
+                          y=self.y,
+                          sensitive_features=self.sensitive_features)
+
+        df, sensitive_feature_name, label_name = credo_data._concat_features_label_to_dataframe()
+        results = {}
+
+        # Distribution of samples across groups
+        sb = df.groupby([sensitive_feature_name]).agg(count=(label_name, len), percentage=(label_name, lambda x: 100.0*len(x)/len(df))).reset_index().to_dict(orient='records')
```
Comment (IanAtCredo) [inline credoai/modules/dataset_modules/dataset_base.py:181]: nit: make multiline

Comment (IanAtCredo) [reply credoai/modules/dataset_modules/dataset_base.py:181]: nit: `sb` is not immediately understandable

File: credoai/modules/dataset_modules/dataset_base.py [hunk 5]
```diff
@@ -156,6 +159,44 @@ def _calculate_mutual_information(self, categorical_features=None, normalize=Tru
...
+        results['sample_balance'] = sb
+
+        # Distribution of samples across groups
+        lb = df.groupby([sensitive_feature_name, label_name]).size().unstack(fill_value=0).stack().reset_index(name='count').to_dict(orient='records')
+        results['label_balance'] = lb
```
Comment (IanAtCredo) [inline credoai/modules/dataset_modules/dataset_base.py:186]: nit: `lb` is not the immediately understandable

**Review by IanAtCredo — CHANGES_REQUESTED** [review]
The `static_method` change is the main one. Having `CredoModel` and `CredoData` defined in a module is very anti-pattern

File: credoai/modules/dataset_modules/dataset_base.py (hunk 4 above)
Comment (amrasekh) [reply credoai/modules/dataset_modules/dataset_base.py]: applied

File: credoai/modules/dataset_modules/dataset_base.py (hunk 5 above)
Comment (amrasekh) [reply credoai/modules/dataset_modules/dataset_base.py]: applied

**Review by IanAtCredo — APPROVED** [review]

**Pull request merged**